### PR TITLE
Profile

### PR DESCRIPTION
**Pull request by mb-dbc-dk**
The ability to extract OpenAgency profiles into query expressions

This is two parts.

1 From web-api a parser of BadgerFish responses (as OpenAgency provides)
2 The parsing of OpenAgency responses, and ability to join profiles

### CODE REVIEW

File: solr/src/main/java/dk/dbc/opensearch/solr/Solr.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.regex.Pattern;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.impl.CloudSolrClient;
+import org.apache.solr.client.solrj.impl.HttpSolrClient;
+
+/**
+ *
+ * @author DBC {@literal <dbc.dk>}
+ */
+public class Solr {
```
Comment (mran-dbc) [inline solr/src/main/java/dk/dbc/opensearch/solr/Solr.java:36]: _Comment suggestion:_
Helper class to build a Solr Http Client from an URL
URL pattern should be on the form: ZK://<hosts>/<root>/<Default collection>
The Hosts/Root part is optional
Hosts can be a list of hostnames separated by , (comma)

File: solr/src/main/java/dk/dbc/opensearch/solr/Solr.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    private static final Pattern ZK = Pattern.compile("zk://([^/]*)(/.*)?/([^/]*)");
```
Comment (mran-dbc) [inline solr/src/main/java/dk/dbc/opensearch/solr/Solr.java]: Hvis min regexp Fu er rigtig er det muligt at lave en URL uden hverken hosts eller root delen -- Er det som det skal være?
group 2 fra regex'en indeholder den foranstillede slash -- Er det som det skal være?

File: solr/src/main/java/dk/dbc/opensearch/solr/SolrQueryFields.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ * opensearch-solr is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package dk.dbc.opensearch.solr;
```
Comment (mran-dbc) [inline solr/src/main/java/dk/dbc/opensearch/solr/SolrQueryFields.java:19]: _Comment suggestion:_
Contains a prepared Solr query based on the given set of Solr-rules, Profile and cql.

The Solr query is flattened and split into a set of nested queries and filter queries for better performance.
The profile, query and its parts is stored for later caching.

File: solr/src/main/java/dk/dbc/opensearch/solr/profile/Profile.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ *
+ * @author DBC {@literal <dbc.dk>}
+ */
+public class Profile  implements Serializable {
```
Comment (mran-dbc) [inline solr/src/main/java/dk/dbc/opensearch/solr/profile/Profile.java]: _Comment suggestion:_
An Agency profile fetched from OpenAgency.
Holds the Solr search- and relation filerqueries and the set of relations to search (_Or is it relations allowed?_).

File: solr/src/main/java/dk/dbc/opensearch/solr/profile/Profiles.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ *
+ * @author DBC {@literal <dbc.dk>}
+ */
+public class Profiles implements Serializable {
```
Comment (mran-dbc) [inline solr/src/main/java/dk/dbc/opensearch/solr/profile/Profiles.java:59]: _(Comment suggestion)_
A collection of OpenAgency profiles to be used for a request.

A number og OpenAgency profiles, can be given in the request, and must be held for further processing.

**Review by mran-dbc — APPROVED** [review]
Har lavet forslag til nogle kommentarer

**Review by petergarnaes — APPROVED** [review]

**Pull request merged**